aten/src/ATen/native/cuda/ForeachUtils.cuh

```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include <ATen/ATen.h>
+#include <ATen/native/cuda/Loops.cuh>
+#include <ATen/native/cuda/MemoryAccess.cuh>
+namespace at { 
+namespace native {
+namespace {
+
+static constexpr int64_t kILP = 4;
+static constexpr int64_t kChunkSize = 65536;
+static constexpr int64_t kBlockSize = 512;
+
+template<typename T>
+__device__ __forceinline__ bool is_aligned(T* p){
+  return ((uint64_t)p) % (kILP * sizeof(T)) == 0;
+}
+
+template<typename T>
+__device__ __forceinline__ void load_store(T* dst, T* src, int dst_offset, int src_offset){
+  using LT = at::native::memory::aligned_vector<T, kILP>;
+  ((LT*)dst)[dst_offset] = ((LT*)src)[src_offset];
+}
+
+}
+
+bool check_fast_route(TensorList tensors, Scalar scalar) {
+  TORCH_CHECK(tensors.size() > 0, "Tensor list must have at least one tensor.");
+  auto expected_dtype = tensors[0].dtype();
+  auto expected_strides = tensors[0].strides();
+
+  for (auto t : tensors) {
+    if (t.dtype() != expected_dtype) {
+      return false;
+    }
+
+    if (t.strides() != expected_strides) {
+      return false;
+    }
+
+    if (t.layout() != at::kStrided) {
+      return false;
+    }
+
+    if (!t.is_non_overlapping_and_dense()) {
+      return false;
+    }
+
+    if (at::isIntegralType(t.scalar_type()) != scalar.isIntegral()) {
+     return false;
+    }
```

Comment:
@ngimel, im really not sure about this check. We are trying to avoid type promotion issue but im worried that this will cause a lot of calls end up on a slow path. Is that alright?

Reply:
I think there are 2 things you want to avoid:
  integral tensor + floating scalar
  boolean tensor (because boolean tensor + int scalar is promoted to int64 as you've discovered)
That way, a common situation of `float list +1` will still go to fast path.
Also, you can take this check out of the loop and check it for only one of the tensors.   
And please cover it with tests if you did not already.

Reply:
Covered this and added test `test_add_scalar_with_different_scalar_type` which covers all the cases: 
- int tensor with float scalar
- float tensor with int scalar
- bool tensor with int scalar
- bool tensor with float scalar